src/test/java/part2/exercise/CollectorsExercise2.java

```diff
@@ -168,34 +166,57 @@ public void collectKeyValueMap() {
         final List<Pair> pairs = generatePairs(10, 100);
 
         // В два прохода
-        // final Map<String, Key> keyMap1 = pairs.stream()...
+        final Map<String, Key> keyMap1 = pairs.stream()
+                .map(Pair::getKey)
+                .collect(Collectors.toMap(Key::getId,Function.identity(),(key1,key2)->key1));
 
-        // final Map<String, List<Value>> valuesMap1 = pairs.stream()...
+        final Map<String, Set<Value>> valuesMap1 = pairs.stream()
+                .collect(Collectors.groupingBy(
+                        pair -> pair.getValue().getKeyId(),
+                        Collectors.mapping(Pair::getValue, Collectors.toSet())));
 
         // В каждом Map.Entry id ключа должно совпадать с keyId для каждого значения в списке
-        // final Map<Key, List<Value>> keyValuesMap1 = valueMap1.entrySet().stream()...
+        final Map<Key, Set<Value>> keyValuesMap1 = combineMaps(keyMap1, valuesMap1);
 
         // В 1 проход в 2 Map с использованием MapPair и mapMerger
         final MapPair res2 = pairs.stream()
                 .collect(new Collector<Pair, MapPair, MapPair>() {
                     @Override
                     public Supplier<MapPair> supplier() {
-                        // TODO
-                        throw new UnsupportedOperationException();
+                        return MapPair::new;
                     }
 
                     @Override
                     public BiConsumer<MapPair, Pair> accumulator() {
-                        // TODO add key and value to maps
-                        throw new UnsupportedOperationException();
+                        return (mapPair, pair) -> {
+                            mapPair.getKeyById().merge(pair.getKey().getId(), pair.getKey(), (k1, k2) -> k1);
+                            mapPair.getValueById().merge(
+                                    pair.getValue().getKeyId(),
+                                    new HashSet<>(Collections.singletonList(pair.getValue())),

```

Comment:
Avoid useless `HashSet` creation. Use `computeIfAbsent` + `add` instead of `merge`